enterprise/coderd/identityprovider/middleware.go

```diff
@@ -0,0 +1,129 @@
+package identityprovider
+
+import (
+	"net/http"
+	"net/url"
+
+	"github.com/coder/coder/v2/coderd/httpapi"
+	"github.com/coder/coder/v2/coderd/httpmw"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/site"
+)
+
+// authorizeMW serves to remove some code from the primary authorize handler.
+// It decides when to show the html allow page, and when to just continue.
+// TODO: For now only browser-based auth flow is officially supported but in a
+// future PR we should support a cURL-based flow.
+func authorizeMW(accessURL *url.URL) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			origin := r.Header.Get(httpmw.OriginHeader)
+			originU, err := url.Parse(origin)
+			if err != nil {
+				httpapi.Write(r.Context(), rw, http.StatusBadRequest, codersdk.Response{
+					Message: "Invalid origin header.",
+					Detail:  err.Error(),
+				})
+				return
+			}
+
+			referer := r.Referer()
+			refererU, err := url.Parse(referer)
+			if err != nil {
+				httpapi.Write(r.Context(), rw, http.StatusBadRequest, codersdk.Response{
+					Message: "Invalid referer header.",
+					Detail:  err.Error(),
+				})
+				return
+			}
+
+			app := httpmw.OAuth2ProviderApp(r)
+			ua := httpmw.UserAuthorization(r)
+
+			// url.Parse() allows empty URLs, which is fine because the origin is not
+			// always set by browsers (or other tools like cURL).  If the origin does
+			// exist, we will make sure it matches.  We require `referer` to be set at
+			// a minimum, however.
+			cameFromSelf := (origin == "" || originU.Hostname() == accessURL.Hostname()) &&
+				refererU.Hostname() == accessURL.Hostname() &&
+				refererU.Path == "/login/oauth2/authorize"
```

Comment:
Oh yeah that does sound like a nightmare.  Hmmm...so `referer` is only required to detect the `allow` part of it, and it could be blank on the initial HTML render if opened from a desktop app or something I think.  So maybe I error if both redirected is set and referer is blank, since the redirected param indicates the user might have pressed the button.

There is a todo for cURL at the top of the function!  But I should move it down next to the HTML part.

Reply:
> So maybe I error if both redirected is set and referer is blank, since the redirected param indicates the user might have pressed the button.

This is perfect. I was thinking the exact same thing on why it would be difficult. But this solution is perfect for now imo.

We had a customer that bitbucket had this error because their proxy was nuking the Referrer header. It was the hardest thing in the world to debug.